### ISSUE TITLE
Inaccurate (pessimistic?) AUC estimates

### ISSUE THREAD

**Issue by turambar** [deeplearning4j/deeplearning4j#3516]
#### Issue Description

Hey all, I've noticed that the DL4J area under the ROC curve (AUC) estimates -- for binary problems, at the very least -- sometimes disagree with those of [scikit-learn](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html). In my admittedly non-exhaustive comparisons, I would say that when they disagree, DL4J typically appears to be underestimating the AUC (by as much as 50% in extreme cases). I'll try to provide examples in the next few days.

I think I've identified two sources for the discrepancies:

1. For candidate thresholds, DL4J uses a [pre-defined set evenly spaced between 0 and 1](https://github.com/deeplearning4j/deeplearning4j/blob/5240cd90a7ebb7efe62f4887757f867e1ad3106a/deeplearning4j-nn/src/main/java/org/deeplearning4j/eval/ROCBinary.java#L48), while I believe that scikit-learn uses the predicted probabilities in sorted order. The advantage of the fixed steps is that is easier to parallelize, but it can suffer if the step size is larger than the distance between neighboring predicted probabilities. One could simply increase the number of steps / decrease the step size, but this quickly becomes infeasible for small probabilities, e.g., points from two different classes differ by, say, 1E-7 (you need millions of thresholds).

2. The relationship between FPR and TPR can be one-to-many. Our [current approach to computing AUC](https://github.com/deeplearning4j/deeplearning4j/blob/5240cd90a7ebb7efe62f4887757f867e1ad3106a/deeplearning4j-nn/src/main/java/org/deeplearning4j/eval/ROCBinary.java#L357) does not take this into account, at least not in the same way that scikit-learn appears to. We treat each point separately, which can lead to underestimating AUC w.r.t. scikit-learn, which I believe uses only the highest TPR value for each value of FPR.

#### Version Information

Please indicate relevant versions, including, if relevant:

DL4J/ND4J/etc.: 0.8.1-SNAPSHOT

#### Contributing

If you'd like to help us fix the issue by contributing some code, but would
like guidance or help in doing so, please mention it!

**Comment by turambar** [comment]
@AlexDBlack

**Comment by turambar** [comment]
@huitseeker you might be interested in this as well.

**Comment by AlexDBlack** [comment]
I've done some reading and thinking on this topic, looks like there's a few options:

1. Simple thresholding (i.e., what we do currently)
2. Exact (requires all prediction probabilities in memory)
3. Binned (approximate; lower memory requirements that exact, but more accurate than simple thresholding)
4. Incremental on-line (exact and approximate)

**Exact**
Exact is reasonably straightforward to implement; the main issues are (a) memory requirements (and hence serialization/saving performance etc), and (b) computational cost.

Memory requirements: for each output, we need to store predictions and class labels. Assume a double (8 bytes) and a binary label (1 bit); ignoring the negligible binary label cost, let's assume 8 bytes per example, *per output* (multi-class, and multiple binary output cases are multiplied by number of outputs).
Best case: 8MB per 1M examples for memory. The main computational cost is sorting (n log(n)) so I expect that we would be able scale an exact implementation to a few million examples without performance degrading too much.

**Binning**
From what I understand, Spark's MLLib implementation uses binning:
https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/evaluation/BinaryClassificationMetrics.scala#L26-L42
this differs from our approach by setting thresholds dynamically such that we have an equal number per bin (should work much better than what we currently have for skewed probabilities).
Downside: I imagine we require the entire dataset available at the time of evaluation (Spark's implementation seems to assume this); i.e., we can't do it online? Therefore, not an option here.

**Incremental Online Approaches**

The following link provides some good information on the subject: http://www.cs.waikato.ac.nz/~remco/roc.pdf
Exact incremental AUC is possible; apparently this is even more efficient than the naive exact implementation. However, implementation is more complex and memory requirements appear to be larger. And notably, we'd have to update/review the algorithm to work out how to merge separate partial calculations (for example, for distributed) which would not be easy.

--------------------------

Proposal:
- Add the option to do exact AUROC/AUCPR
- ROC: default to thresholded, but maybe increase the default number of thresholds to say 50 or 100
- If exact calculation performance becomes a problem, revisit binning/incremental approaches

**Comment by turambar** [comment]
@AlexDBlack sounds like a reasonable plan. Is "exact" a variant of the sorted probabilities approach? Is it possible to do that in parallel on Spark, or would we have to "collect" all the probabilities?

**Comment by AlexDBlack** [comment]
@turambar exact = "sorted probabilities", yes
The implementation I'm working on will store predictions + classes in an INDArray then do a sortRows op when the user finally calls calculateAUC() (and caching the result, obviously).
Upside of that is we'll get native c++ / GPU sorting too (once @raver119 does some optimizations there, which we have discussed), so it should scale OK I think. But I'll run some benchmarks on it at some point.

**Comment by AlexDBlack** [comment]
https://github.com/deeplearning4j/deeplearning4j/pull/3524

Use exact calculation by setting numSteps = 0.
Exact implementation was validated vs. scikit learn.
